workflow-service/src/main/java/com/redhat/parodos/workflow/execution/aspect/WorkFlowExecutionAspect.java

```diff
@@ -105,53 +82,23 @@ public WorkReport executeAroundAdvice(ProceedingJoinPoint proceedingJoinPoint, W
 
 		/* get workflow definition entity */
 		WorkFlowDefinition workFlowDefinition = this.workFlowDefinitionRepository.findFirstByName(workflowName);
-
-		String masterWorkflowName = WorkContextDelegate.read(workContext,
-				WorkContextDelegate.ProcessType.WORKFLOW_DEFINITION, WorkContextDelegate.Resource.NAME).toString();
-		boolean isMaster = workflowName.equals(masterWorkflowName);
-
-		WorkFlowExecution workFlowExecution;
-		WorkFlowExecution masterWorkFlowExecution = null;
-		String arguments = WorkFlowDTOUtil.writeObjectValueAsString(
-				WorkContextDelegate.read(workContext, WorkContextDelegate.ProcessType.WORKFLOW_EXECUTION, workflowName,
-						WorkContextDelegate.Resource.ARGUMENTS));
-		UUID projectId = UUID.fromString(WorkContextDelegate
-				.read(workContext, WorkContextDelegate.ProcessType.PROJECT, WorkContextDelegate.Resource.ID)
-				.toString());
-
-		// get master WorkFlowExecution, this is the first time execution for master
-		// workflow if return null
-		UUID masterWorkFlowExecutionId = Optional.ofNullable(WorkContextDelegate.read(workContext,
-				WorkContextDelegate.ProcessType.WORKFLOW_EXECUTION, WorkContextDelegate.Resource.ID))
-				.map(id -> UUID.fromString(id.toString())).orElse(null);
-
-		if (masterWorkFlowExecutionId == null) {
-			workFlowExecution = handleFirstTimeMainWorkFlowExecution(projectId, workFlowDefinition, arguments,
-					workContext);
-		}
-		else {
-			masterWorkFlowExecution = workFlowRepository.findById(masterWorkFlowExecutionId)
-					.orElseThrow(() -> new WorkflowExecutionNotFoundException(
-							"masterWorkFlow not found for sub-workflow: " + workflowName));
-
-			// get the workflow execution if this is triggered by continuation service
-			WorkFlowExecution finalMasterWorkFlowExecution = masterWorkFlowExecution;
-			workFlowExecution = isMaster ? masterWorkFlowExecution
-					: Optional
-							.ofNullable(workFlowRepository.findFirstByWorkFlowDefinitionIdAndMasterWorkFlowExecution(
-									workFlowDefinition.getId(), masterWorkFlowExecution))
-							.orElseGet(() -> this.workFlowService.saveWorkFlow(projectId, workFlowDefinition.getId(),
-									WorkFlowStatus.IN_PROGRESS, finalMasterWorkFlowExecution, arguments));
-
-			if (workFlowExecution.getStatus().equals(WorkFlowStatus.COMPLETED)) {
-				// skip the workflow if it is already successful
-				if (workFlowDefinition.getType().equals(WorkFlowType.CHECKER)) {
-					workFlowSchedulerService.stop((WorkFlow) proceedingJoinPoint.getTarget());
-				}
-				return new DefaultWorkReport(WorkStatus.COMPLETED, workContext);
+		WorkFlowExecutionInterceptor executionHandler = workFlowExecutionFactory
+				.createExecutionHandler(workFlowDefinition, workContext);
+		WorkFlowExecution workFlowExecution = executionHandler.handlePreWorkFlowExecution();
+
+		// @Richard, I think this is the right place to check if the workflow is already
```

Comment:
@RichardW98 pls let me know what do you think of this

Reply:
if masterWorkFlowExecutionId is null, it means this is the first execution of master workflow, and it is not completed for sure

Reply:
@RichardW98  in that case, 
```
if (workFlowExecution.getStatus().equals(WorkFlowStatus.COMPLETED)) {
...
```
is false and it will not be executed.
My question is should this condition be moved to `MasterWorkFlowExecutionInterceptor` or can it be relevant for any Completed workflow?

Reply:
it is relevant to all completed workflow